lib/rules/no-unused-prop-types.js

```diff
@@ -725,6 +713,60 @@ module.exports = {
     }
 
     /**
+     * Marks all props found inside ObjectTypeAnnotaiton as declared.
+     * 
+     * Modifies the declaredProperties object
+     * @param {ASTNode} propTypes 
+     * @param {Object} declaredPropTypes 
+     * @returns {Boolean} True if propTypes should be ignored (e.g. when a type can't be resolved, when it is imported)
+     */
+    function declarePropTypesForObjectTypeAnnotation(propTypes, declaredPropTypes) {
+      let ignorePropsValidation = false;
+
+      iterateProperties(propTypes.properties, (key, value) => {
+        if (!value) {
+          ignorePropsValidation = true;
+          return;
+        }
+
+        const types = buildTypeAnnotationDeclarationTypes(value, key);
+        types.fullName = key;
+        types.name = key;
+        types.node = value;
+        declaredPropTypes.push(types);
+      });
+
+      return ignorePropsValidation;
+    }
+
+    /**
+     * Marks all props found inside IntersectionTypeAnnotation as declared.
+     * Since InterSectionTypeAnnotations can be nested, this handles recursively. 
+     * 
+     * Modifies the declaredPropTypes object
+     * @param {ASTNode} propTypes 
+     * @param {Object} declaredPropTypes 
+     * @returns {Boolean} True if propTypes should be ignored (e.g. when a type can't be resolved, when it is imported)
+     */
+    function declarePropTypesForIntersectionTypeAnnotation(propTypes, declaredPropTypes) {
```

Comment:
Can these two functions live somewhere they can be shared across the rules that currently duplicate them?

Reply:
Yes. I'm working on it 😄 The idea will be that the rule will be declared as:

```
create: propTypes.detect((context, components, utils) => {
...
})
```

And this will take care of ALL the shared code between `prop-types` and `no-unused-prop-types`. Perhaps also for the other rules.